serde-generate/src/dart.rs

```diff
@@ -0,0 +1,1283 @@
+            writeln!(
+                &mut emitter.out,
+                "import '../{0}/{0}.dart';",
+                encoding.name()
+            )?;
+        }
+
+        if let Some(files) = &self.config.external_definitions.get("import") {
+            for file in *files {
+                writeln!(&mut emitter.out, "import '{0}';", file)?;
+            }
+        }
+
+        writeln!(&mut emitter.out, "\nexport '../serde/serde.dart';")?;
+
+        writeln!(&mut emitter.out, "\npart 'trait_helpers.dart';")?;
+        for name in registry.keys() {
+            writeln!(&mut emitter.out, "part '{}.dart';", name.to_snake_case())?;
+        }
+
+        Ok(())
+    }
+
+    fn write_container_class(
+        &self,
+        dir_path: &std::path::Path,
+        current_namespace: Vec<String>,
+        name: &str,
+        format: &ContainerFormat,
+        registry: &Registry,
+    ) -> Result<()> {
+        let mut file =
+            std::fs::File::create(dir_path.join(name.to_string().to_snake_case() + ".dart"))?;
+        let mut emitter = DartEmitter {
+            out: IndentedWriter::new(&mut file, IndentConfig::Space(2)),
+            generator: self,
+            current_namespace,
+            registry,
+        };
+
+        emitter.output_preamble()?;
+        emitter.output_container(name, format)
+    }
+
+    fn write_helper_class(
+        &self,
+        dir_path: &std::path::Path,
+        current_namespace: Vec<String>,
+        registry: &Registry,
+    ) -> Result<()> {
+        let mut file = std::fs::File::create(dir_path.join("trait_helpers.dart"))?;
+        let mut emitter = DartEmitter {
+            out: IndentedWriter::new(&mut file, IndentConfig::Space(2)),
+            generator: self,
+            current_namespace,
+            registry,
+        };
+
+        emitter.output_preamble()?;
+        emitter.output_trait_helpers(registry)
+    }
+}
+
+impl<'a, T> DartEmitter<'a, T>
+where
+    T: Write,
+{
+    fn output_preamble(&mut self) -> Result<()> {
+        writeln!(
+            self.out,
+            "part of {}_types;",
+            self.generator.config.module_name
+        )?;
+
+        Ok(())
+    }
+
+    fn get_field_container_type(&self, name: &str) -> Option<&ContainerFormat> {
+        match self.registry.get(name) {
+            Some(container) => Some(container),
+            None => None,
+        }
+    }
+
+    // in Dart enums cannot have a static method added to them
+    // yet so we must call the extension class instead
+    fn get_class(&self, name: &str) -> String {
+        if self.generator.config.c_style_enums {
+            use ContainerFormat::Enum;
+            match self.get_field_container_type(name) {
+                Some(Enum(_)) => format!("{}Extension", name),
+                _ => name.to_string(),
+            }
+        } else {
+            name.to_string()
+        }
+    }
+
+    fn quote_qualified_name(&self, name: &str) -> String {
```

Comment:
So until now, we have not attempted to remap conflicting names on the fly with a list like this in each language. (I feel like the list may always be growing)

If we must, perhaps we can make this behavior a codegen configuration option with a default value that is documented for each language.

@dimroc what do you think?